pennylane/clusters.py

```diff
@@ -0,0 +1,476 @@
+# Copyright 2018-2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains high-level QNode processing functions and classes.
+"""
+# pylint: disable=too-many-arguments,import-outside-toplevel
+from collections.abc import Sequence
+
+from pennylane.qnodes import QNode
+from pennylane.measure import expval, var, sample
+from pennylane.operation import Observable
+
+
+MEASURE_MAP = {"expval": expval, "var": var, "sample": sample}
+
+
+def map(template, observables, device, measure="expval", interface="autograd", diff_method="best"):
+    """Map a quantum template over a list of observables to create
+    a :class:`QNodeCluster`.
+
+    The number of QNodes within the created QNode cluster will match the number
+    of observables passed. The device and the measurement type will either be
+    applied to all QNodes in the cluster, or can be provided as a list for more
+    fine-grained control.
+
+    Note that the template **must** have the following signature:
+
+    .. code-block:: python
+
+        template(params, wires, **kwargs)
+
+    where
+
+    * ``params`` are the trainable weights of the variational circuit,
+    * ``wires`` is a list of integers representing the wires of the device, and
+    * ``kwargs`` are any additional keyword arguments that need to be passed
+      to the template.
```

Comment:
It might be nice to have the example right after this section or include a ``See example below`` disclaimer (though this might not be too specific). This would give the idea of how exactly the ``template`` can be created without the user having to scroll (or even think that an example was not provided).

Reply:
Yeah, I was unsure about this. @mariaschuld and I have discussed an approach where the examples occur _below_ the arugment list, like the NumPy docs: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.linalg.eigh.html

I think the best solution is to make the argument list shorter?

Reply:
I've made the argument list shorter, and simply link to the qnode decorator for more details.

Reply:
But I agree, I think we should add a 'Example ⇓' button using CSS to all pages :)